File: Zag-Smalltalk/zig/fib-hand.zig
const std = @import("std");
const config = @import("zag/config.zig");
const tailCall = config.tailCall;
const trace = config.trace;
const stdCall = config.stdCall;
const object = @import("zag/zobject.zig");
const Object = object.Object;
const Context = @import("zag/context.zig").Context;
const execute = @import("zag/execute.zig");
const PC = execute.PC;
const SP = execute.SP;
const TFProcess = execute.TFProcess;
const tfAsProcess = execute.tfAsProcess;
const TFContext = execute.TFContext;
const tfAsContext = execute.tfAsContext;
const MethodSignature = execute.MethodSignature;
const compileMethod = execute.compileMethod;
const Sym = @import("zag/symbol.zig").symbols;
// fib
//   self <= 2 ifTrue: [^ 1]
//   ^ (self - 1) fib + (self - 2) fib
const _fibCM = compileMethod(Sym.fibonacci,0,2,.SmallInteger,.{&fib,&fib1,&fib2});
const fibCM = _fibCM.asCompiledMethodPtr();
export fn fib(pc: PC, sp: SP, _process: TFProcess, _context: TFContext, _: MethodSignature) callconv(stdCall) SP {
    const process = tfAsProcess(_process);
    const context = tfAsContext(_context);
    const self = sp.top;
    if (self.rawU() <= Object.from(2).rawU()) {
        sp.top = Object.from(1);
        return @call(tailCall, context.npc, .{ context.tpc, sp, process, context, undefined});
    }
    const ctxt = @call(.never_tail, &Context.push, .{context, sp, process, fibCM, 0, 2, 1});
    const newSp = ctxt.asNewSp();
    const newerSp = newSp.pushRawInt(self.rawU()-Object.from(1).untaggedInt());
    ctxt.setReturnBoth(fib1, pc.next());
    return @call(tailCall, &fib, .{ fibCM.codePc(), newerSp, process, ctxt, undefined});
}
pub fn fib1(pc: PC, sp: SP, _process: TFProcess, _context: TFContext, _: MethodSignature) callconv(stdCall) SP {
    const process = tfAsProcess(_process);
    const context = tfAsContext(_context);
    const self = context.getSelf();
    const newSp = sp.pushRawInt(self.rawU()-Object.from(2).untaggedInt());
    context.setReturnBoth(fib2, pc.next());
    return @call(tailCall, &fib, .{ fibCM.codePc(), newSp, process, context, undefined});
}
pub fn fib2(_: PC, sp: SP, _process: TFProcess, _context: TFContext, _: MethodSignature) callconv(stdCall) SP {
    const process = tfAsProcess(_process);
    const context = tfAsContext(_context);
    const result = sp.top.rawU()+sp.next.untaggedInt();
    const spAndContext = @call(.never_tail,&Context.pop, .{context,process});
    const newSp = spAndContext.sp;
    newSp.top = @bitCast(result);
    const ctxt = spAndContext.ctxt;
    return @call(tailCall, ctxt.npc, .{ ctxt.tpc, newSp, process, ctxt, undefined});
}
test "nothing" {
    _ = &fib;
}

File: Zag-Smalltalk/zig/fib-hand.c
// fib
//   self <= 2 ifTrue: [^ 1]
//   ^ (self - 1) fib + (self - 2) fib
typedef unsigned long long u64;
typedef long long i64;
typedef short u16;
typedef union { u64 u; char a[8];} Object;
#define from(i) (cast(i + 0xFFFC000000000000LL))
#define cast(v) ((Object) {.u = v})
#define untaggedInt(o) (o.u - 0xFFFC000000000000LL)
typedef u64 HeapHeader;
#define header(x) (x)
#define contextHeaderOnStack(x) ((u64)119<<48|(x)<<16)
typedef struct CompiledMethod CompiledMethod;
typedef struct { Object top, next, third, fourth; } *SP;
typedef struct {unsigned int hash;unsigned short class;} MethodSignature;
#define ms(h, c) ((MethodSignature){.hash = h,.class = c})
MethodSignature noMS;
typedef struct Process Process;
typedef struct Context Context;
typedef union Code *PC;
typedef SP ThreadedFn(PC pc, SP sp, Process *process, Context *context, MethodSignature signature);
typedef union Code {ThreadedFn *prim; i64 i; u64 u; Object object;} Code;
#define stack_size 909
struct Process {
  Object stack[stack_size];
  Process *next;
  u64 id;
  u64 trapContextNumber;
  // and more that we don't care about
};
struct Context {
  HeapHeader header;
  CompiledMethod *method; // note this is not an Object, so access and GC need to handle specially
  PC tpc; // threaded PC
  ThreadedFn *npc; // native PC - in Continuation Passing Style
  Context *prevCtxt; // note this is not an Object, so access and GC need to handle specially
  u64 trapContextNumber;
  Object temps[1];
};
#define Context_Basesize 6
struct CompiledMethod {
  HeapHeader header;
  Object stackStructure;
  MethodSignature signature;
  ThreadedFn *verifier;
  Code code[3];
};
extern SP verifyMethod(PC, SP, Process*, Context *, MethodSignature);
SP fib(PC, SP, Process *, Context *, MethodSignature);
SP fib1(PC, SP, Process *, Context *, MethodSignature);
SP fib2(PC, SP, Process *, Context *, MethodSignature);
CompiledMethod fibCM = {header(0),from(0),ms(0,0),verifyMethod,{&fib,&fib1,&fib2}};
extern Context *push(Context*, SP, Process*, CompiledMethod*, u16, u16, u16);
static int isOnStack(Context *self) {
  return (self->header>>48&15)==0;
}
SP fib(PC pc, SP sp, Process *process, Context *context, MethodSignature signature) {
  Object self = sp->top;
  if (self.u <= from(2).u) {
    sp->top = from(1);
    return context->npc(context->tpc, sp, process, context, noMS);
  }
  Context *newContext = (Context*)(sp - Context_Basesize);
  if ((SP)newContext-2>=(SP)process) {
    newContext->header = contextHeaderOnStack(Context_Basesize + 0);
    newContext->method = &fibCM;
    newContext->prevCtxt = context;
    newContext->trapContextNumber = ((Process*)((u64)process&~0x1F))->trapContextNumber;
    context = newContext;
  } else {
    context = push(context, sp, process, &fibCM, 0, 2, 1);
  }
  sp = ((SP)context)-1;
  sp->top = cast(self.u-untaggedInt(from(1)));
  context->npc = &fib1;
  context->tpc = pc+1;
  return fib(pc-1,sp,process,context,signature);
}
SP fib1(PC pc, SP sp, Process *process, Context *context, MethodSignature signature) {
  Object self = context->temps[0];
  SP newSp = sp-1;
  newSp->top = cast(self.u-untaggedInt(from(2)));
  context->npc = &fib2;
  context->tpc = pc+1;
  return fib(pc-2,newSp,process,context,signature);
}
typedef struct {SP sp;Context *ctxt;} SPC;
SPC pop(Context*,Process*);
SP fib2(PC pc, SP sp, Process *process, Context *context, MethodSignature signature) {
  Object result = cast(sp->top.u+untaggedInt(sp->next));
  if (isOnStack(context)) {
      sp = ((SP)context)+Context_Basesize;
      context = context->prevCtxt;
    } else {
      SPC spAndContext = pop(context,process);
      sp = spAndContext.sp;
      context = spAndContext.ctxt;
    }
  sp->top = result;
  return context->npc( context->tpc, sp, process, context, signature);
}
